### ISSUE TITLE
Error when upgrading to version 3.6.3 from 3.6.2 

### ISSUE THREAD

**Issue by JosieBigler** [httptoolkit/mockttp#142]
After upgrading to 3.6.3 from 3.6.2 I get an error with some of my requests that have a large(ish) File in the body of the http request. 

I will update this issue when I nail down, exactly what's going on.  Looking at some of the changes, I'm not terribly surprised.  

Some pointers or help on how to go about debugging would be great. 

I'm using node version 18.13.0

My server code 

```
    server.forAnyRequest()
        .forHostname('url.com')
        .thenForwardTo(`${Url}`, {
            beforeRequest: (request) => {
                if(request.path != "/health-check"){
                    console.log(request);
                }
                request.url = `${Url}` + request.path;
                request.headers.host = Url;
                return request;
            }
        });
        
        server.start(config.port);
```

The error node throws: 

```
node:events:491
      throw er; // Unhandled 'error' event
      ^

Error: aborted
    at connResetException (node:internal/errors:718:14)
    at abortIncoming (node:_http_server:747:17)
    at socketOnClose (node:_http_server:741:3)
    at TLSSocket.emit (node:events:525:35)
    at node:net:320:12
    at Socket.done (node:_tls_wrap:588:7)
    at Object.onceWrapper (node:events:628:26)
    at Socket.emit (node:events:525:35)
    at TCP.<anonymous> (node:net:320:12)
Emitted 'error' event on PassThrough instance at:
    at emitErrorNT (node:internal/streams/destroy:151:8)
    at emitErrorCloseNT (node:internal/streams/destroy:116:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  code: 'ECONNRESET'
```

I've been trying to nail down exactly which commit, my hunch is it's one of these two: 

https://github.com/httptoolkit/mockttp/commit/c3e65b42d62a6cb9a9fc282e94e32c83c5c8c7ed
https://github.com/httptoolkit/mockttp/commit/fb452dd7c7ef5bf8ca092eb1492ced673b532aea

Update. Debugging, the line that throws is 
https://github.com/httptoolkit/mockttp/blob/82520698f55e82417423097f8acbc61d2bddc024/src/util/buffer-utils.ts#L37

**Comment by JosieBigler** [comment]
Debugging further, the application seems to hang on this line. 

https://github.com/httptoolkit/mockttp/blob/82520698f55e82417423097f8acbc61d2bddc024/src/util/request-utils.ts#L296

**Comment by JosieBigler** [comment]
The size of the file is something like ~20MB.  We've handled up to... 200MB+ in the previous version.

**Comment by pimterry** [comment]
Hmm, it's hard to know exactly what's causing this. Do you see the same issue on the latest version? (v3.7.4)

One thing I can see immediately is that you're using URLs and hostnames interchangeably, which is not correct, and could plausibly cause strange issues. I'm specifically talking about `request.url`, which should be a fully specified URL like `https://example.com/abc`, and `headers.host` which should _never_ be a URL, it should just be a plain hostname like `example.com`.

The `thenForwardTo` parameter will accept hostnames with or without a protocol (if you don't specify, it preserves the protocol from the original request) but `request.url` and host headers have a specific format they expect.

Otherwise I'm not sure what could cause this, and in general I know it's certainly possible to stream responses large significantly than 20MB via Mockttp with no problem (definitely at least in the latest version, and every other recent version AFAIK) so there must be something specific about the configuration here that's triggering this...

Would it be possible to create a standalone minimal repro for this, so I can test it directly and investigate myself?

**Comment by JosieBigler** [comment]
Ok, here is a simple reproduction of the error.   Startup the server and then send a POST with a file as form-data.  I attached a simple PDF that was ~800KB.  

https://github.com/JosieBigler/mockttperror

Screenshot of the Postman request: 

![image](https://user-images.githubusercontent.com/34112636/236629440-b148e7b7-bd1c-4e11-a0c5-ef74884ed8ae.png)
![image](https://user-images.githubusercontent.com/34112636/236629460-cc8d5c9e-f744-4171-90eb-bb3a811d07af.png)

**Comment by pimterry** [comment]
After quite a bit of digging, I think I know what caused that error - there was a possible deadlock for request bodies if they hit an internal buffer limit before running `beforeRequest` in proxying scenarios like this.

I've just pushed a commit for a potential solution with a different internal streaming approach that _should_ fix this. Can you try checking out the latest changes from `main` and see if this now works correctly for you?

**Comment by JosieBigler** [comment]
Hi Tim, working as expected now with our setup :)

**Comment by pimterry** [comment]
This is now released as Mockttp v3.7.5. Thanks for reporting this! Glad we could get that sorted :+1:
